core/src/subgraph/loader.rs

```diff
@@ -4,207 +4,71 @@ use std::time::Instant;
-        &self,
-        unresolved_data_sources: Vec<UnresolvedDataSource>,
-        logger: &Logger,
-    ) -> Result<Vec<DataSource>, Error> {
-        // Resolve the data sources and return them
-        let mut result = Vec::new();
-        for item in unresolved_data_sources.into_iter() {
-            let resolved = item.resolve(&*self.link_resolver, logger).await?;
-            result.push(resolved);
-        }
-        Ok(result)
+    pub fn new(store: Arc<S>) -> Self {
+        Self { store }
     }
 }
 
 #[async_trait]
-impl<L, Q, S> DataSourceLoaderTrait for DataSourceLoader<L, Q, S>
+impl<S> DataSourceLoaderTrait for DataSourceLoader<S>
 where
-    L: LinkResolver,
-    Q: GraphQlRunner,
     S: Store + SubgraphDeploymentStore,
 {
     async fn load_dynamic_data_sources(
         &self,
         deployment_id: SubgraphDeploymentId,
         logger: Logger,
+        manifest: SubgraphManifest,
     ) -> Result<Vec<DataSource>, Error> {
         let start_time = Instant::now();
 
+        let template_map: HashMap<&str, &DataSourceTemplate> = HashMap::from_iter(
+            manifest
+                .templates
+                .iter()
+                .map(|template| (template.name.as_str(), template)),
+        );
         let mut data_sources = vec![];
 
-        loop {
-            let skip = data_sources.len() as i32;
-            let query = self.dynamic_data_sources_query(&deployment_id, skip)?;
-            let query_result = self
-                .graphql_runner
-                .cheap_clone()
-                .query_metadata(query)
-                .await?;
-            let unresolved_data_sources =
-                self.parse_data_sources(&deployment_id, query_result.as_ref().clone())?;
-            let next_data_sources = self
-                .resolve_data_sources(unresolved_data_sources, &logger)
-                .await?;
-
-            if next_data_sources.is_empty() {
-                break;
-            }
-
-            data_sources.extend(next_data_sources);
+        for stored in self.store.load_dynamic_data_sources(&deployment_id)? {
+            let StoredDynamicDataSource {
+                name,
+                source,
+                context,
+            } = stored;
+
+            let template = template_map.get(name.as_str()).ok_or_else(|| {
+                format_err!(
+                    "deployment `{}` does not have a template called `{}`",
+                    deployment_id.as_str(),
+                    name
+                )
+            })?;
+            let context = context
+                .map(|ctx| serde_json::from_str::<Entity>(&ctx))
+                .transpose()?;
+
+            let ds = DataSource {
+                kind: template.kind.clone(),
+                network: template.network.clone(),
+                name,
+                source,
+                mapping: template.mapping.clone(),
+                context,
+                templates: Vec::new(),
```

Comment:
If we're doing this we should go ahead and remove the `templates` field from `DataSource`. Some old subgraphs might break, but now is our last opportunity to do this sort of breakage, so I'm in favor of going ahead.

Reply:
There are 5 subgraphs that currently have templates in `ethereum_contract_data_source`:
```
name         | raisehq/raisekovan
name         | gvillarc91/gaston
name         | perfectmak/marketprotocol
name         | perfectmak/marketprotocol-kovan
name         | kohshiba/forecastory-mainnet
```
We probably need to talk to their owners to get them to redeploy.

Reply:
This PR already breaks them, which I'm fine with. If you don't want to remove the templates field right now, could you put a comment above the field declaration it should be removed?